oodeel/datasets/torch_data_handler.py

```diff
@@ -0,0 +1,485 @@
+        self._check_keys_and_dataset()
+
+    def _check_keys_and_dataset(self):
+        """Check if the initialization arguments are correct"""
+        assert isinstance(
+            self._dataset[0], (Tuple, List, torch.Tensor)
+        ), "Dataset to be wrapped needs to output tuple of tensors"
+        assert len(self._dataset[0]) == len(
+            self._raw_output_keys
+        ), "Length mismatch between tuple of tensors and dictionary keys"
+
+    @property
+    def output_keys(self):
+        return self._raw_output_keys + self._added_output_keys
+
+    @property
+    def output_shapes(self):
+        dummy_item = self[0]
+        return [dummy_item[key].shape for key in self.output_keys]
+
+    def __getitem__(self, index):
+        """Return a dictionary of tensors corresponding to a specfic index"""
+        tensors = self._dataset[index]
+        output_dict = {
+            key: tensor for (key, tensor) in zip(self._raw_output_keys, tensors)
+        }
+        for map_fn in self.map_fns:
+            output_dict = map_fn(output_dict)
+        return output_dict
+
+    def map(self, map_fn: Callable, inplace: bool = False):
+        """Map the dataset
+
+        Args:
+            map_fn (Callable): map function f: dict -> dict
+            inplace (bool): if False, applies the mapping on a copied version of\
+                the dataset. Defaults to False.
+
+        Return:
+            DictDataset: Mapped dataset
+        """
+        dataset = self if inplace else copy.deepcopy(self)
+        dataset.map_fns.append(map_fn)
+        return dataset
+
+    def filter(self, filter_fn: Callable, inplace: bool = False):
+        """Filter the dataset
+
+        Args:
+            filter_fn (Callable): filter function f: dict -> bool
+            inplace (bool): if False, applies the filtering on a copied version of\
+                the dataset. Defaults to False.
+
+        Returns:
+            DictDataset: Filtered dataset
+        """
+        indices = [
+            i
+            for i in tqdm(range(len(self)), desc="Filtering the dataset...")
+            if filter_fn(self[i])
+        ]
+        dataset = self if inplace else copy.deepcopy(self)
+        dataset._dataset = Subset(self._dataset, indices)
+        return dataset
+
+    def concatenate(self, other_dataset: Dataset, inplace: bool = False):
+        """Concatenate with another dataset
+
+        !!! warning
+            The map functions that will be retained will be those of this dataset
+
+        Args:
+            other_dataset (Dataset): Dataset to concatenate with
+            inplace (bool): if False, applies the filtering on a copied version of\
+                the dataset. Defaults to False.
+
+        Returns:
+            DictDataset: Concatenated dataset
+        """
+        dataset = self if inplace else copy.deepcopy(self)
+        dataset._dataset = ConcatDataset([self._dataset, other_dataset])
+        return dataset
```

Comment:
Do you really need to implement these methods? Couldn't you put the code directly in the corresponding methods of `TorchDataHandler`, namely `map_ds`, `filter_by_feature_value` and `merge`? In addition to reduce the size of the code, IMO it could improve readability for pytorch users that would immediately see familiar code under the mentioned methods without having to refer to the custom DictDataset. Moreover, it feels a bit like a wrapper in a wrapper. 

With that said, the aim is for this section of the code to be easily readable for pytorch users, so since I am not a regular pytorch user, I may be wrong.

Reply:
Actually, these methods are really specific to the `DictDataset` class as they require to get access and modify some of its private attributes. So I think keeping them inside the `DictDataset` would be more suitable? Morever, some torch users could possibly appreciate to get access to these functionalities without passing by the DataHandler (as they can do with the tensorflow dataset class).

Reply:
Ok it makes sense to me, thank you.